Add EIP-5994: Token Balance Update Hooks (Pods) (#6000)
ethereum/EIPs: EIPS/eip-5994.md

---
eip: 5994
title: Token Balance Update Hooks (Pods)

> description: Extends EIP-20, EIP-721, EIP-1155 token standards to enable opt-in balance tracking by external contracts (Pods)

Reviewer: ```suggestion
description: Extends EIP-20, EIP-721, and EIP-1155 to enable opt-in balance tracking by external contracts (Pods)
```

---
eip: 5994
title: Token Pods extensions (ERC20/ERC721)

> description: Token extension enabling external smart contract based Pods to track balances of those users who opted-in to these Pods.

Author: Added mentioning EIP-1155:
> description: Extends EIP-20, EIP-721, EIP-1155 token standards to enable opt-in balance tracking by external contracts (Pods)

author: Anton Bukov (@k06a), Mikhail Melnik (@zumzoom)
discussions-to: https://ethereum-magicians.org/t/eip-6000-token-pods-extensions-erc20-erc721/11782
status: Draft
type: Standards Track
category: ERC
created: 2022-11-18

> requires: 20, 721

Author: In short: there is no way to restrict which exact token_id is supported by the Pod with the current Pod API.

> requires: 20, 721

Reviewer: So, why not add it? This is what draft EIPs are for: so that the API can be changed without risking breaking existing applications?

> requires: 20, 721

Author: Finally added EIP-1155 support

---

## Abstract

> Token Pods concept introduces token extensions for the most popular token standards: ERC20, ERC721 and ERC1155, which allows holders of assets to connect their balances to Pods, which could track changes of their balances.

Author: Will remove EIP-1155 for now

## Motivation

> This Token Pods concept allows to have permissionless extensions for tokens, which can be used for incentivisations (aka farms) or delegations or any other ideas.

Author: Updated motivation to the following:
> This concept of Pods allows to extend the logic of tokens in the most permissionless and risk-free way. Almost arbitrary logical extensions are possible: incentives for farmers, delegation of rights, voting, etc.

Example implementation can be found here:
- https://github.com/1inch/erc20-pods

Examples of Pods:
- https://github.com/1inch/farming
- https://github.com/1inch/delegating

## Specification

> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.

Author: Agree, will use uppercase for important words

}

interface IERC20Pods is IERC20 {
    function hasPod(address account, address pod) external view returns(bool);
    function podsCount(address account) external view returns(uint256);
    function podAt(address account, uint256 index) external view returns(address);
    function pods(address account) external view returns(address[] memory);
    function podBalanceOf(address pod, address account) external view returns(uint256);

    function addPod(address pod) external;
    function removePod(address pod) external;
    function removeAllPods() external;
}

interface IERC721Pods is IERC721 {
    function hasPod(address account, address pod) external view returns(bool);
    function podsCount(address account) external view returns(uint256);
    function podAt(address account, uint256 index) external view returns(address);
    function pods(address account) external view returns(address[] memory);
    function podBalanceOf(address pod, address account) external view returns(uint256);

    function addPod(address pod) external;
    function removePod(address pod) external;
    function removeAllPods() external;

> }

Reviewer: This should be ideally be split into three sections, each with NatSpec. In addition, I think that pods, in order to be versatile enough for the use cases mentioned in the rationale, should:

* Have a hook for `addPod` that allows the pod the chance to revert and update tracking variables
* Have a hook for `removePod` that allows the pod the chance to revert and update tracking variables
* **Have separate hooks and hook names for EIP-20, EIP-721, and EIP-1155**. EIP-20 and EIP-1155 pods should take an amount, EIP-721 pods shouldn't. And EIP-721 and EIP-1155 should take a tokenId, and EIP-20 shouldn't.
* For EIP-721 and EIP-1155, pods should be able to be added per token ID or for the token as a whole.
* (This is a big one for me) EIP-165 should be supported.

Also, where is the EIP-1155 interface?

```

### Pod specification

> - Pod MUST implement interface `IPod` and fit execution of `updateBalances` method in 200,000 gas and never revert.

Reviewer 2: ```suggestion
- Pod MUST implement interface `IPod` and fit execution of `updateBalances` method within 200,000 gas and never revert.
```

For `in`, do you mean `within`?

Could you provide rationale about how the number `200,000` was  picked??

Author: @xinbenlv number is relatively big for any possible logic, but small enough to have limitation 5-10 pods per account to fit into more sophisticated transaction as part of token transfer. We expect in practice users will have up to 2-3 pods with 30-50k gas consumption max for each, in average nearly 10-20k each.

- Pod SHOULD make sure `updateBalances(...)` method was triggered by `token` it was designated for.

- Pod can track changes of balances of users who voluntarily added this pod, all other account will be represented as `address(0)` and interactions with such accounts could be considered as mints and burns.

> - Pod SHOULD NOT try to change balances of the token it was designated for, since it would introduce reentrancy, which will not pass.

Reviewer: I recommend moving this to natspec.

But regardless, the functions defined in the token extensions are not specified, which is a problem.

## Rationale

> 

Reviewer: This is missing. A placeholder is acceptable:

```suggestion

Needs discussion.

```

## Backwards Compatibility

> Fully backward compatibile with ERC20 and ERC721.

Reviewer: ```suggestion
Fully backward compatible with EIP-20, EIP-721, and EIP-1155.
```
There was a spelling mistake, and EIP-1155 was missing. Also, as previously mentioned, the format for EIPs is currently `EIP-X`, where `X` is the EIP's number.

(This might change. If it does, I'll change it myself)

## Security Considerations

> Pods implementations will never block balance update, but can eat upto 200,000 gas. Number of added Pods per account is limited to some immutable value (reasonable amount taking into account max transaction gas limit).

Reviewer: * ERCs MUST NOT depend on gas costs. They change from one hard fork to the next. EIP-165 is the one exception to this as it was finalized before that rule was put in place. It's well-recognized that the 30,000 gas limit is not a "real" rule, however.
* This isn't in the spec!

Author: I see no way to avoid this constant. This number should be specified. Moreover we have restriction of number of pods added to one account. N*200k should be less than max transaction gas limit. So I expect devs to use 200k constant and 10-20 pods limit per account.

Author: Core concept is permissionless pods and risk-free participation. So Pods cant block transfers/mint/burns of designated assets.

Reviewer: It's okay to have the gas limit as a should. But it is *not* okay to have it as a must.

> Pods implementations will never block balance update, but can eat upto 200,000 gas. Number of added Pods per account is limited to some immutable value (reasonable amount taking into account max transaction gas limit).

Reviewer 2: I am a little worry about the introduction of Pods and its hooks will massively increase the chance of renetry attack, could you provide your considerations about this?
